### ISSUE TITLE
SC 2.4.11 how to interpret adjacent contrast when the indicator is both in and out of the control?

### ISSUE THREAD

**Issue by mraccess77** [w3c/wcag#2300]
What happens when you have a default Chrome 2px indicator that is 1px outside the component and 1px on the border of the component?
* Size passes
* Focused/unfocused state - requires testing
* Adjacent colors - do you only test the 1px that is outside the control?  Since the indicator as a whole is 2px - is this not applicable even though 1px of the indicator is not adjacent?

Other tangential notes:
From what I can tell the default focus indicator in Chrome is 2 CSS pixels thick for a button.  The default border also appears to be 2px thick.   Focusing a button replaces the default border with the indicator. 

* Size passes
* unfocused and focused states may fail depending on the border and indicator color
* Adjacent colors - not applicable

If you have a 1 pixel border and the default 2px focus indicator - chrome seems to replace the border and also takes an interior rectangle of pixels as well.  The same case goes when there is no border - it put the 2px indicator where the border should be and takes an internal rectangle as well.
* Size passes
* Unfocused and focused states may fail depending on the border and indicator color
* Adjacent colors - not applicable

**Comment by alastc** [comment]
> a default Chrome 2px indicator

In that case it passes due to being the default, but I'll proceed on the assumption that an author is over-riding it and creating a similar indicator.

Working with the [latest language](https://raw.githack.com/w3c/wcag/focus-appearance-encompass/understanding/22/focus-appearance-minimum.html), "An area of the focus indicator meets all the following". (NB: This is one of the cases that was discussed in the [examples document](https://docs.google.com/document/d/1TI_DsJjfg9RW_A9C1XfqgtfYqNMfnxeCz6lxjpOQ9rk/edit#heading=h.tcutgbbczlhr), 1st example in the table)

So an area has to meet all bullets, but the 1px that is overlapping the border doesn't meet change-of-contrast. So you are left with a 1px outline around the component, which doesn't meet adjacent contrast because it's the same color.

I'm not sure about your other notes, if the above comment doesn't answer your question, could you take a focused/unfocused screenshot?

**Comment by mraccess77** [comment]
Thanks @alastc I had no idea the SC had changed so much - I'd like to know more information about the indicator's background is not changed exception - does that apply both when an indicator is within the component and (separate situation) if it's outside of the component? 

Back to my original question though - I think what you are saying is if the border was black and now the indicator is black we only look at the change and compare what changed to the adjacent button's black border?  Even though technically the black on the border is set by the indicator we don't care for this as visually it didn't appear as a visual change and so we are still treating it as not part of the indicator.

**Comment by mbgower** [comment]
> I think what you are saying is if the border was black and now the indicator is black we only look at the change and compare what changed to the adjacent button's black border? Even though technically the black on the border is set by the indicator we don't care for this as visually it didn't appear as a visual change and so we are still treating it as not part of the indicator.

Correct. The definition is:
> Focus indicator
pixels that are changed to visually indicate when a user interface component is in a focused state

The assessment is on the pixels changed.

**Comment by alastc** [comment]
> about the indicator's background is not changed exception - does that apply both when an indicator is within the component and (separate situation) if it's outside of the component?

Yes, that's why it says "The focus indicator and the indicator's background color are not modified by the author". One of the examples I picked up was changing the background of a button in Edge (and I assume Chrome) impacts the default indicator. 

So the wording is to keep in scope author-changes that affect either the indicator, or the background of the indicator, where ever that is.

**Comment by alastc** [comment]
Just a little thought on the language, where it has "the indicator's background color", the word "background" is one I've learned to avoid for this SC.

Would it be better as?
"The focus indicator, and the same pixels in the unfocused state, are not modified by the author, or"

Hmm, I don't think that's an improvement, I'll keep thinking.

**Comment by mraccess77** [comment]
Thank you both - this clears up my question.  I can close this issue unless someone tells me to keep it open.

**Comment by mbgower** [comment]
@mraccess77 Just as a post-mortem, I will add additional guidance in the Understanding document around the "background". Originally I had page background, which was changed since it didn't capture in the inset focus indicator (e.g., inside a button) scenario you mentioned.

**Comment by mbgower** [comment]
@alastc I've added a spaceholder for pairs of indicator backgrounds to demonstrate the two common use cases, including the one you mention in https://github.com/w3c/wcag/issues/2300#issuecomment-1105779261

Could you please add in those images?

**Comment by mraccess77** [comment]
Based on discussion today on the call - it would be worth making sure we agree that if the indicator is two-toned - do you need to compare the focus indicator to itself?

**Comment by mbgower** [comment]
I'm going to swap in the definition in place of _focus indicator_ in the preamble, which will make it easier for this discussion:
> When a user interface component has keyboard focus, the pixels that are changed to visually indicate when a user interface component is in a focused state:

So point 1, we're assessing the pixels changed.

Here are the 3 primary considerations:
>- Encloses the visual presentation of the user interface component;
 >- Has a contrast ratio of at least 3:1 between its pixels in the focused and unfocused states;
 >- Has a contrast ratio of at least 3:1 against adjacent colors.

Second point: two-toned indicators  are rarely (ever?) going pass the second of those primary checks because by design some of their pixels are not going to contrast sufficiently against page background, so we're always going to be thinking about the exception wording. So let's go there...

---

Here are the sub-bullets of the exception:
>An area of the focus indicator meets all the following:
>- Is at least as large as the area of a 1 CSS pixel thick perimeter of the unfocused component...;
>- Has a contrast ratio of at least 3:1 against the same pixels in the unfocused state;
>- Has a contrast ratio of at least 3:1 against adjacent colors, or is no thinner than 2 CSS pixels.

I'm tackling one at a time:
> Is at least as large as the area of a 1 CSS pixel thick perimeter

I've truncated it to make the key point here: a certain area of the pixels of the 2-tone focus indicator need to be big enough. Assuming you use 2 solid linesfor your contrasting lines and they are outside the indicator, either is going to meet that. Okay?

> 3:1 against the same pixels in the unfocused state;

For the second sub-bullet, let's say for the sake of argument, one of your focus rectangles is very light and the other is very dark, which is essentially the second example in [Creating a two-color focus indicator...](https://www.w3.org/WAI/WCAG22/Techniques/css/C40):
![image](https://user-images.githubusercontent.com/14298245/165438337-57ac78ba-6e3a-49c2-b29e-58be87e5b3ab.png)
Depending on what the page background is (e.g., white or dark-mode), the pixels making up one of those rectangles are going to meet the second bullet, okay?

> Has a contrast ratio of at least 3:1 against adjacent colors, or is no thinner than 2 CSS pixels.

Assuming you abut the 2 contrasting lines, they are always going to meet contrast on one side, right? Then, for either a light or dark background, one of those two lines is also going to contrast on the other side too. 

Make sense?

**Comment by mraccess77** [comment]
I like that the 2nd option (former exception) allows for perimeter rather than the "encloses" language as this would allow the control border to be used as the indicator.   Is that your read as well?  Encloses to me would mean something outside the border.

**Comment by mbgower** [comment]
>  allows for perimeter rather than the "encloses" language as this would allow the control border to be used as the indicator. Is that your read as well?

Yes, and remember as well that it doesn't have to _be_ the perimeter. It's just using perimeter to establish the size of the area of the focus indicator. A focus indicator that meets the overall area can be inside a button, for instance. Note that where the border was used as the indicator, it will still need to meet all the other subbullets (for contrast and/or thickness).

> Encloses to me would mean something outside the border.

"encloses" is defined as "Solidly bounds or surrounds". So if there is a border to a button, the focus indicator would have to be outside the border to enclose it (though technically it can abut it, with no space between).

Just to make sure you're aware, the latest revamped version alters the normative text layout (but not the meaning) https://raw.githack.com/w3c/wcag/focus-appearance-structure/understanding/22/focus-appearance.html
